src/library.js

```diff
@@ -3412,6 +3434,61 @@ LibraryManager.library = {
     }
   },
 
+#if SHRINK_LEVEL == 0 && !RELOCATABLE
+  // A mirror copy of contents of wasmTable in JS side, to avoid relatively
+  // slow wasmTable.get() call. Only used when not compiling with -Os or -Oz.
+  // TODO: This code currently assumes that the elements added to the wasm table are
+  // never removed, so cannot be enabled in RELOCATABLE builds.
+  $wasmTableMirror: [],
+
+  $setWasmTableEntry: function(idx, func) {
+    wasmTable.set(idx, func);
+    wasmTableMirror[idx] = func;
+  },
+
+  $wbind__deps: ['$wasmTableMirror'],
+  $wbind: function(funcPtr) {
+    var func = wasmTableMirror[funcPtr];
+    if (!func) {
+      if (funcPtr >= wasmTableMirror.length) wasmTableMirror.length = funcPtr + 1;
+      wasmTableMirror[funcPtr] = func = wasmTable.get(funcPtr);
+    }
+#if ASSERTIONS
+    assert(wasmTable.get(funcPtr) == func, "JavaScript-side Wasm function table mirror is out of date!");
+#endif
+    return func;
+  },
+
+#else
+
+  $setWasmTableEntry: function(idx, func) {
+    wasmTable.set(idx, func);
+  },
+
+  $wbind: function(funcPtr) {
+    // In -Os and -Oz builds, do not implement a JS side wasm table mirror for small
+    // code size, but directly access wasmTable, which is a bit slower as uncached.
+    return wasmTable.get(funcPtr);
+  },
+#endif
+
+  // A helper that binds a wasm function into a form that can be called by passing all
+  // the parameters in an array, e.g. wbindArray(func)([param1, param2, ..., paramN]).
+  $wbindArray__deps: ['$wbind'],
```

Comment:
Give that this function has exactly one caller internally (`createInvokeFunction` in `library_dylink.js`) ... I wonder if we can avoid creating a new public API here and just create `wbind`.

I think this is especially true if you consider that this `SHRINK_LEVEL` optimization should probably be moved into that function where it could be even more valuable (since it would avoid the Arguments.slice)?  

We can always add `wbindArray` later.. but removing it once it added is harder..

Reply:
That works, dropped wbindArray.